mne/beamformer/_lcmv.py

```diff
@@ -39,27 +50,38 @@ def _reg_pinv(x, reg):
     return linalg.pinv(x), d
 
 
-def _setup_picks(picks, info, forward, noise_cov=None):
-    if picks is None:
-        picks = pick_types(info, meg=True, eeg=True, ref_meg=False,
-                           exclude='bads')
+def _setup_picks(info, forward, data_cov=None, noise_cov=None):
+    # get a list of all channel names:
+    fwd_ch_names = set([c['ch_name'] for c in forward['info']['chs']])
 
-    ok_ch_names = set([c['ch_name'] for c in forward['info']['chs']])
-    if noise_cov is not None:
-        ok_ch_names.union(set(noise_cov.ch_names))
+    # handle channels from forward model and info:
+    ch_names = [c['ch_name'] for c in info['chs']
+                if c['ch_name'] not in info['bads'] and
+                c['ch_name'] in fwd_ch_names]
 
-    if noise_cov is not None and set(info['bads']) != set(noise_cov['bads']):
+    # inform about excluding channels:
+    if (data_cov is not None and set(info['bads']) != set(data_cov['bads']) and
+            (set(ch_names) & set(data_cov['bads']))):
+        logger.info('info["bads"] and data_cov["bads"] do not match, '
+                    'excluding bad channels from both.')
+    if (noise_cov is not None and
+            set(info['bads']) != set(noise_cov['bads']) and
+            (set(ch_names) & set(noise_cov['bads']))):
```

Comment:
I'm wondering if we want use all these sets, duplicate names should actually never exist. Unless for the bads someone happens to add the same channel twice. But no strong emotions here.